thorlcr/graph/thgraphmaster.cpp

```diff
@@ -1609,44 +1612,43 @@ void CJobMaster::broadcast(ICommunicator &comm, CMessageBuffer &msg, mptag_t mpt
         replyTag = queryJobChannel(0).queryMPServer().createReplyTag();
         msg.setReplyTag(replyTag);
     }
-    if (globals->getPropBool("@broadcastSendAsync", true)) // only here in case of problems/debugging.
+    auto f = [&](unsigned r) -> bool
     {
-        class CSendAsyncfor : public CAsyncFor
+        if (!aborting || comm.verifyConnection(r, (unsigned)-1, false))
         {
-            CMessageBuffer &msg;
-            mptag_t mptag;
-            unsigned timeout;
-            StringAttr errorMsg;
-            ICommunicator &comm;
-        public:
-            CSendAsyncfor(ICommunicator &_comm, CMessageBuffer &_msg, mptag_t _mptag, unsigned _timeout, const char *_errorMsg)
-                : comm(_comm), msg(_msg), mptag(_mptag), timeout(_timeout), errorMsg(_errorMsg)
-            {
-            }
-            void Do(unsigned i)
-            {
-                if (!comm.send(msg, i+1, mptag, timeout))
-                    throw createBCastException(i+1, errorMsg);
-            }
-        } afor(comm, msg, mptag, timeout, errorMsg);
+            if (!comm.send(msg, r, mptag, timeout))
+                throw createBCastException(r, errorMsg);
+            return true;
+        }
+        return false;
+    };
+    std::vector<std::future<bool>> sendResultsFuture;
+    Owned<IException> sendExcept;
+    for (unsigned dstrank=1; dstrank<=groupSizeExcludingMaster; ++dstrank)
+        sendResultsFuture.push_back(std::async(f, dstrank));
```

Comment:
What is the advantage of switching to std::async and futures?   Because more than one can throw an exception which should not be fatal?

(One reason for asking is the possibility of other changes that would remove the need for starting new threads in CASyncFor)

Reply:
That is the main reason: It is not possible to return a value from the thread, so it wouldn't be able to track the number of failed sends.  If it can't track the number of successful sends, then it would be waiting for replies that will never arrive.  (And that delays the completion of the job for the "timeout" period.)

Also, the code is more compact and standard.

Reply:
I think the overhead (of new threads) here is inconsequential, because the threads are not being used for speed, but for concurrent network waits, but I agree, if we start having a pool that allows reuse for such things that would make sense, and that would be a good reason to use our own constructs.

However, I do think the lambda's and capture (and to some extend the return pattern/futures) make the code clearer and more readable. If we were to have a pooling mechanism I think it should support lambda functions..

Reply:
The asyncFor functions support lambdas.